### PR TITLE
Add minion transport failover

### PR DESCRIPTION
**Pull request by techhat**
### What does this PR do?

Allows a minion's transport option to be set to `detect`, which will attempt to connect to the master first using `tcp`, and then using `zeromq`.
### What issues does this PR fix or reference?
#33881
### Tests written?

No

### CODE REVIEW

**Comment by cachedout** [comment]
I do think this will work but I'd like @DmitryKuzmenko and @thatch45 to look at it as well.

File: salt/minion.py [hunk 1]
```diff
@@ -587,8 +587,21 @@ def eval_master(self,
                 opts.update(prep_ip_port(opts))
                 opts.update(resolve_dns(opts))
                 try:
-                    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
-                    yield pub_channel.connect()
+                    if self.opts['transport'] == 'detect':
+                        detect_connect = False
+                        self.opts['detect_mode'] = True
```
Comment (DmitryKuzmenko) [inline salt/minion.py]: Am I right that `detect_mode` option is a flag meaning _"currently we're detecting the actual transport"_. Shouldn't it be set to `False` or unset after we're successfully connected?

File: salt/minion.py [hunk 2]
```diff
@@ -587,8 +587,21 @@ def eval_master(self,
...
+                        for trans in ('tcp', 'zeromq'):
+                            if detect_connect is True:
+                                continue
+                            self.opts['transport'] = trans
+                            pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
+                            yield pub_channel.connect()
+                            if not pub_channel.auth.authenticated:
+                                continue
+                            detect_connect = True
```
Comment (DmitryKuzmenko) [inline salt/minion.py]: The `detect_connect` flag looks redundant. The code will be better readable if you'll just break if `authenticated`.

Comment (cachedout) [inline salt/minion.py]: +1 I had this same thought but didn't bring it up. Thanks for mentioning this, @DmitryKuzmenko

File: salt/transport/zeromq.py [hunk 3]
```diff
@@ -380,8 +380,9 @@ def __del__(self):
     def connect(self):
         if not self.auth.authenticated:
             yield self.auth.authenticate()
-        self.publish_port = self.auth.creds['publish_port']
-        self._socket.connect(self.master_pub)
+        if self.opts.get('detect_mode', False) is False:
+            self.publish_port = self.auth.creds['publish_port']
+            self._socket.connect(self.master_pub)
```
Comment (DmitryKuzmenko) [inline salt/transport/zeromq.py]: Maybe I'm wrong, sorry if so. But I don't understand what this change for? What if we entered here in "detect" mode and `authenticate` call succeeded. In this case the channel is "authenticated" but it's socket is disconnected. Do you plan connect it later?

File: salt/minion.py (hunk 1 above)
Comment (techhat) [inline salt/minion.py]: `detect_mode` means "`transport` was originally set to `detect`, but we can't leave it like that if we're trying to connect, but we still need to remember that, so let's store it somewhere else".

File: salt/transport/zeromq.py (hunk 3 above)
Comment (techhat) [inline salt/transport/zeromq.py]: Without this change, if the master is set to `zeromq` and the minion is set to detect, you will get this error:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado/ioloop.py", line 604, in _run_callback
    ret = callback()
  File "/usr/lib/python2.7/site-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/tornado/ioloop.py", line 619, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 237, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 1024, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/site-packages/salt/crypt.py", line 505, in _authenticate
    self._authenticate_future.set_exception(error)
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 283, in set_exception
    getattr(exception, '__traceback__', None)))
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 306, in set_exc_info
    self._set_done()
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 320, in _set_done
    for cb in self._callbacks:
TypeError: 'NoneType' object is not iterable
```
The error does not appear if the master is set to `tcp`.

File: salt/minion.py (hunk 2 above)
Comment (techhat) [inline salt/minion.py]: Fixed.

**Comment by thatch45** [comment]
This should work because we set the transport in the opts dict early enough in the minion startup that all subsequent calls to it will result int he correct transport mechanism. I was also heavily involved in creating this.

I think that we should set the default to detect and run the test test suite, then if the test suite passes take the default back to zeromq before merging.

Are there any legitimate test failures?

**Comment by thatch45** [comment]
So long story short, Joseph and I spent a lot of time working to find the best place to set the detection, and this is the entry point for transport in all minion startup cases, and this verifies the transport in what we feel is a very natural way.

**Comment by jtand** [comment]
Go Go Jenkins!

File: salt/transport/zeromq.py (hunk 3 above)
Comment (techhat) [inline salt/transport/zeromq.py]: @DmitryKuzmenko this error is now showing up in the tests either way. Do you have any thoughts on where we can look?

Comment (DmitryKuzmenko) [inline salt/transport/zeromq.py]: @techhat it looks some future has been set done twice.

**Comment by damon-atkins** [comment]
Cannot see anything in the changes to detect if zeromq library is available, and if not go straight to tcp?  So that zeromq can become an optional dependency.

**Comment by thatch45** [comment]
@damon-atkins has a good point here

**Comment by techhat** [comment]
With these last two commits, the `NoneType` error should be taken care of, and `zeromq` is gated.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by damon-atkins** [comment]
Bit more feed back
* Can not see anything which prints error "please install zeromq python library" and dies if configuration file transport is set to zeromq and zeromq is not available. i.e. HAS_ZMQ is False
* if HAS_ZMQ is True import salt.utils.zeromq otherwise don't

**Comment by damon-atkins** [comment]
Link to "Remove zeromq package dependency in Carbon release" #35724

**Comment by techhat** [comment]
Looks like I've got at least one `NoneType` error out of the way, but we're not clear yet. `salt-call` works, but a running minion won't talk to the master.

**Comment by techhat** [comment]
Go Go Jenkins!

**Comment by DmitryKuzmenko** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
Will this pass tests if we change the default transport back to zeromq?
We don't want to change the default until Nitrogen, if they pass tests with the default set to zeromq then we can merge this and then focus on getting the default setting to detect in Nitrogen.

**Comment by techhat** [comment]
Changed the default back, to see if it still works.

**Comment by thatch45** [comment]
Good call, I think that he detect order should stay in sync with the relative release's default for the master

**Pull request merged**